app/src/chat_engine.py

```diff
@@ -0,0 +1,113 @@
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Optional, Sequence
+
+import src.adapters.db as db
+from src.db.models.document import Chunk
+from src.format import format_guru_cards
+from src.generate import generate
+from src.retrieve import retrieve
+from src.shared import get_embedding_model
+from src.util.class_utils import all_subclasses
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class OnMessageResult:
+    response: str
+    chunks: Sequence[Chunk]
+    data: Optional[dict] = None
+
+
+class ChatEngineInterface(ABC):
+    engine_id: str
+    name: str
+
+    @abstractmethod
+    async def on_start(self) -> dict:
+        pass
```

Comment:
Should this really return None?

```suggestion
    @abstractmethod
    async def on_start(self) -> None:
        pass
```

Reply:
Yea, I thought about that too. Removing...